internal/util/logging/mongo_handler.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"maps"
+	"runtime"
+	"slices"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/FerretDB/FerretDB/v2/internal/util/must"
+)
+
+// mongoHandler is a [slog.Handler] that writes logs by using mongo structured JSON format.
+// The format returns log entries with Relaxed Extended JSON specification.
+// The format is not stable.
+//
+//nolint:vet // for readability
+type mongoHandler struct {
+	opts *NewHandlerOpts
+
+	ga        []groupOrAttrs
+	testAttrs map[string]any
+
+	m   *sync.Mutex
+	out io.Writer
+}
+
+// MongoLog represents a single log message in mongo structured JSON format.
+// For now it lacks the fields that are not used by any caller.
+//
+//nolint:vet // to preserve field ordering
+type MongoLog struct {
```

Comment:
Could we think of a better name? Even `MongoLogRecord` (similar to `slog.Record`) is probably better.